### ISSUE TITLE
Investigate adding a macro that gets a free port number

### ISSUE THREAD

**Issue by mlorbetske** [dotnet/templating#152]
Like the "now" macro, we should add a "port" macro with the following characteristics:
- A port number that is not in use by the system at the time of template instantiation should be the result
- Two port number generations cannot have the same result for the same instantiation operation
- The port numbers should be able to be bounded (low/high value)

**Comment by mlorbetske** [comment]
We should be able to do something like this to get our port number, we'll still need to wrap this up in a macro though

``` C#
    private static int GetRandomFreePort(HashSet<int> alreadyReservedPorts, int lowPort = 5000, int highPort = 65536)
    {
        HashSet<int> takenPorts = new HashSet<int>();

        IPGlobalProperties globalProperties = IPGlobalProperties.GetIPGlobalProperties();
        if (globalProperties != null)
        {
            foreach (TcpConnectionInformation info in globalProperties.GetActiveTcpConnections())
            {
                takenPorts.Add(info.LocalEndPoint.Port);
            }

            foreach (IPEndPoint info in globalProperties.GetActiveTcpListeners())
            {
                takenPorts.Add(info.Port);
            }

            foreach (IPEndPoint info in globalProperties.GetActiveUdpListeners())
            {
                takenPorts.Add(info.Port);
            }
        }

        for (int i = lowPort; i <= highPort; ++i)
        {
            if((alreadyReservedPorts == null || !alreadyReservedPorts.Contains(i)) && !takenPorts.Contains(i))
            {
                return i;
            }
        }

        return -1;
    }
```

**Comment by RehanSaeed** [comment]
This is some code I use to get random **free** port numbers. It knows the ports are free by trying to open them. It also supports getting ports for HTTP and HTTPS (HTTPS is within the range 44300 to 44399.

https://github.com/ASP-NET-MVC-Boilerplate/Templates/blob/master/Source/Boilerplate.FeatureSelection/Services/PortService.cs

**Comment by peterblazejewicz** [comment]
I was pleased with Core using conventions: `5000` port by default, debugger with different default port. I use other tools/templates and I know there are: 3000, 4200 or 5000 ports set by default with options like `-p --port` to change this at runtime. Hey, the 44300 is in fact a VS tooling convention printed in books (`By default, IIS Express uses the 44300 port to publish over SSL`) and blogged: http://www.hanselman.com/blog/WorkingWithSSLAtDevelopmentTimeIsEasierWithIISExpress.aspx

**Comment by mlorbetske** [comment]
The range portion of this should be fixed with #1087
